### ISSUE TITLE
关于多阶段 SFT

### ISSUE THREAD

**Issue by David-19940718** [jingyaogong/minimind#161]
作者大大，感谢开源。对于以下第一行所示的流程：

![Image](https://github.com/user-attachments/assets/841c6165-d4af-42d2-b0e5-76ed63eda171)

这里有个疑问。首先定位到 `train_full_sft.py` 脚本：

```bash
...
def init_model(lm_config):
    tokenizer = AutoTokenizer.from_pretrained('./model/minimind_tokenizer')
    model = MiniMindLM(lm_config)
    moe_path = '_moe' if lm_config.use_moe else ''
    ckp = f'./out/pretrain_{lm_config.dim}{moe_path}.pth'
    state_dict = torch.load(ckp, map_location=args.device)
    model.load_state_dict(state_dict, strict=False)
    Logger(f'LLM总参数量：{sum(p.numel() for p in model.parameters() if p.requires_grad) / 1e6:.3f} 百万')
    model = model.to(args.device)
    return model, tokenizer
...
```

假设我这里通过 `pretrain` 模型训练 `sft_521.jsonl` 数据后；想继续微调 `sft_2048.jsonl`，是不是需要将 `ckp` 改为上一步 SFT 后得到权重？

另外，对于`在推理时通过调整RoPE线性差值，实现免训练长度外推到2048及以上将会很方便。` 请在哪里有涉及到？

期待回复 💇

**Comment by jingyaogong** [comment]
是的，将ckp改成full_sft_xxx.pth；

暂时无需关注推理长度外推，目前没有涉及到，之后会新增一个简单的长度外推说明，放在issue或者discussions，到时会at你；

**Comment by beston123** [comment]
> 是的，将ckp改成full_sft_xxx.pth；
> 
> 暂时无需关注推理长度外推，目前没有涉及到，之后会新增一个简单的长度外推简明教程，放在issue或者discussions，到时会at你；
微调sft_521.jsonl 数据后；想继续微调 sft_2048.jsonl
需要修改 --max_seq_len 2048吗？

**Comment by beston123** [comment]
> 作者大大，感谢开源。对于以下第一行所示的流程：
> 
> ![Image](https://github.com/user-attachments/assets/841c6165-d4af-42d2-b0e5-76ed63eda171)
> 
> 这里有个疑问。首先定位到 `train_full_sft.py` 脚本：
> 
> ...
> def init_model(lm_config):
>     tokenizer = AutoTokenizer.from_pretrained('./model/minimind_tokenizer')
>     model = MiniMindLM(lm_config)
>     moe_path = '_moe' if lm_config.use_moe else ''
>     ckp = f'./out/pretrain_{lm_config.dim}{moe_path}.pth'
>     state_dict = torch.load(ckp, map_location=args.device)
>     model.load_state_dict(state_dict, strict=False)
>     Logger(f'LLM总参数量：{sum(p.numel() for p in model.parameters() if p.requires_grad) / 1e6:.3f} 百万')
>     model = model.to(args.device)
>     return model, tokenizer
> ...
> 假设我这里通过 `pretrain` 模型训练 `sft_521.jsonl` 数据后；想继续微调 `sft_2048.jsonl`，是不是需要将 `ckp` 改为上一步 SFT 后得到权重？
> 
> 另外，对于`在推理时通过调整RoPE线性差值，实现免训练长度外推到2048及以上将会很方便。` 请在哪里有涉及到？
> 
> 期待回复 💇

请问使用 max_seq_len=512预训练和微调sft_521.jsonl 数据后，继续微调 sft_2048.jsonl
需要修改 --max_seq_len 2048吗？ 还是一开始就用2048训练的？

**Comment by jingyaogong** [comment]
> > 是的，将ckp改成full_sft_xxx.pth；
> > 暂时无需关注推理长度外推，目前没有涉及到，之后会新增一个简单的长度外推简明教程，放在issue或者discussions，到时会at你；
> 
> 微调sft_521.jsonl 数据后；想继续微调 sft_2048.jsonl 需要修改 --max_seq_len 2048吗？

readme写了

![Image](https://github.com/user-attachments/assets/b4703769-068e-46a8-87fb-c0cc8023002d)


且，repeat with https://github.com/jingyaogong/minimind/issues/184

**Comment by kk123456789001** [comment]
请问作者我用sft_1024.jsonl做黑盒蒸馏，我看保存时的文件名还是full_sft_{lm_config.dim}{moe_path}.pth，由于没有修改lm_config.dim岂不是最终的保存路径会和之前初步sft的路径重复？感谢作者回答

**Comment by kk123456789001** [comment]
请问作者我用sft_1024.jsonl做黑盒蒸馏，我看保存时的文件名还是full_sft_{lm_config.dim}{moe_path}.pth，由于没有修改lm_config.dim岂不是最终的保存路径会和之前初步sft的路径重复？例如我之前是full_sft_512.pth，但是用1024.jsonl微调后的保存路径还是full_sft_512.pth，而没有变成full_sft_1024.pth吗？感谢作者回答

**Comment by jingyaogong** [comment]
> 请问作者我用sft_1024.jsonl做黑盒蒸馏，我看保存时的文件名还是full_sft_{lm_config.dim}{moe_path}.pth，由于没有修改lm_config.dim岂不是最终的保存路径会和之前初步sft的路径重复？例如我之前是full_sft_512.pth，但是用1024.jsonl微调后的保存路径还是full_sft_512.pth，而没有变成full_sft_1024.pth吗？感谢作者回答
> 

保存成什么很自由，自行修改即可

**Comment by guker** [comment]
![Image](https://github.com/user-attachments/assets/8c4a3cb5-886c-445f-aec6-664cfaff7270)
1. 为什么要先在sft_512.jsonl训练，然后再sft_2048.jsonl训练？
2. 按道理讲，将sft_512.jsonl和sft_2048.jsonl合在一起训练，效果会不会更好？

**Comment by Skye2099** [comment]
> ![Image](https://github.com/user-attachments/assets/8c4a3cb5-886c-445f-aec6-664cfaff7270)
> 
> 1. 为什么要先在sft_512.jsonl训练，然后再sft_2048.jsonl训练？
> 2. 按道理讲，将sft_512.jsonl和sft_2048.jsonl合在一起训练，效果会不会更好？
>    [@jingyaogong](https://github.com/jingyaogong)

![Image](https://github.com/user-attachments/assets/97a37dea-a1a1-457e-8c48-3b7fecd02fcf)

**Comment by sunnysky29** [comment]
@jingyaogong  博主，rlhf 4个模型加载这一套后续会加么，这个流程对于llm 训练理解还是挺关键的
